File: backend/app/models/emotion_diary.py
from sqlalchemy import Column, BigInteger, String, Integer, Text, DateTime, ForeignKey
from sqlalchemy.orm import relationship
from datetime import datetime
from app.config.database import Base
from app.models.emotion_tag import diary_tags


class EmotionDiary(Base):
    """情绪遗产日记模型"""

    __tablename__ = "emotion_diaries"

    id = Column(BigInteger, primary_key=True, index=True, autoincrement=True)
    user_id = Column(BigInteger, ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)

    # 核心字段
    trigger_event = Column(Text, nullable=False)  # 触发事件
    immediate_emotion = Column(String(100), nullable=False)  # 即时情绪
    emotion_intensity = Column(Integer, nullable=False)  # 情绪强度 1-10

    # 可选字段
    physical_reaction = Column(Text, nullable=True)  # 身体反应
    auto_thought = Column(Text, nullable=True)  # 自动化思维
    childhood_memory = Column(Text, nullable=True)  # 童年记忆联结
    pattern_recognition = Column(Text, nullable=True)  # 模式识别

    # 时间戳
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False, index=True)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

    # 关系
    user = relationship("User", back_populates="diaries")
    tags = relationship("EmotionTag", secondary=diary_tags, backref="diaries")
    challenges = relationship("ResponseChallenge", back_populates="diary", cascade="all, delete-orphan")

    def __repr__(self):
        return f"<EmotionDiary(id={self.id}, user_id={self.user_id}, emotion='{self.immediate_emotion}')>"

File: flutter_app/lib/models/emotion_diary.dart
import 'emotion_tag.dart';

class EmotionDiary {
  final int id;
  final int userId;
  final String triggerEvent;
  final String immediateEmotion;
  final int emotionIntensity;
  final String? physicalReaction;
  final String? autoThought;
  final String? childhoodMemory;
  final String? patternRecognition;
  final DateTime createdAt;
  final DateTime updatedAt;
  final List<EmotionTag> tags;

  EmotionDiary({
    required this.id,
    required this.userId,
    required this.triggerEvent,
    required this.immediateEmotion,
    required this.emotionIntensity,
    this.physicalReaction,
    this.autoThought,
    this.childhoodMemory,
    this.patternRecognition,
    required this.createdAt,
    required this.updatedAt,
    this.tags = const [],
  });

  factory EmotionDiary.fromJson(Map<String, dynamic> json) {
    return EmotionDiary(
      id: json['id'] as int,
      userId: json['user_id'] as int,
      triggerEvent: json['trigger_event'] as String,
      immediateEmotion: json['immediate_emotion'] as String,
      emotionIntensity: json['emotion_intensity'] as int,
      physicalReaction: json['physical_reaction'] as String?,
      autoThought: json['auto_thought'] as String?,
      childhoodMemory: json['childhood_memory'] as String?,
      patternRecognition: json['pattern_recognition'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
      tags: (json['tags'] as List<dynamic>?)
              ?.map((tag) => EmotionTag.fromJson(tag as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'trigger_event': triggerEvent,
      'immediate_emotion': immediateEmotion,
      'emotion_intensity': emotionIntensity,
      'physical_reaction': physicalReaction,
      'auto_thought': autoThought,
      'childhood_memory': childhoodMemory,
      'pattern_recognition': patternRecognition,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
      'tags': tags.map((tag) => tag.toJson()).toList(),
    };
  }
}

class DiaryCreateRequest {
  final String triggerEvent;
  final String immediateEmotion;
  final int emotionIntensity;
  final String? physicalReaction;
  final String? autoThought;
  final String? childhoodMemory;
  final String? patternRecognition;
  final List<int> tagIds;

  DiaryCreateRequest({
    required this.triggerEvent,
    required this.immediateEmotion,
    required this.emotionIntensity,
    this.physicalReaction,
    this.autoThought,
    this.childhoodMemory,
    this.patternRecognition,
    this.tagIds = const [],
  });

  Map<String, dynamic> toJson() {
    return {
      'trigger_event': triggerEvent,
      'immediate_emotion': immediateEmotion,
      'emotion_intensity': emotionIntensity,
      'physical_reaction': physicalReaction,
      'auto_thought': autoThought,
      'childhood_memory': childhoodMemory,
      'pattern_recognition': patternRecognition,
      'tag_ids': tagIds,
    };
  }
}
